### PR TITLE
Fix #134 #192 #194. Unified diff colors

### LINKED ISSUE #134: Ediff faces work poorly with dark theme
Using ediff with the `solarized-dark` theme, the backgrounds of diff regions can get quite bright. Combined with dark text (e.g. comments) superimposed, this can make some text very hard to read:

screen shot 2014-11-07 at 2 03 16 pm

Here it is with the light theme for comparison:
screen shot 2014-11-07 at 2 06 12 pm

### LINKED ISSUE #194: Various diff/merge modes are broken
Something entirely new needs to be done for this to work.. 
This is a meta issue.

Related issues: 
- #192 
- #191 
- #134

### PR DESCRIPTION
**Pull request by wyuenho**
Second try to fix #194 based on https://github.com/bbatsov/solarized-emacs/issues/194#issuecomment-395270416

The diff color palate is mostly taken from magit unmodified. The blue for changed is added by myself. ~~All the solarized faces for magit-diff and magit-section that clash with the rest of this new diff palate is removed.~~

Please try this out and tell me what you think.

### CODE REVIEW

File: solarized.el [hunk 1]
```diff
@@ -1412,25 +1370,15 @@ customize the resulting theme."
...
-       ((t (:background ,(solarized-color-blend yellow base03 0.1)))))
-     `(magit-diff-hunk-heading-highlight
-       ((t (:background ,(solarized-color-blend yellow base02 0.1)))))
      `(magit-diff-hunk-heading-selection
        ((t (:background ,(solarized-color-blend yellow base02 0.1)
                         :foreground ,orange
                         :weight bold))))
      `(magit-diff-lines-heading          ((t (:background ,orange
                                                           :foreground ,base3))))
-     `(magit-diff-context-highlight      ((t (:background ,base02))))
```
Comment (wyuenho) [inline solarized.el]: Essentially `s-diff-fine-context-[fb]g`

File: solarized.el [hunk 2]
```diff
@@ -1412,25 +1370,15 @@ customize the resulting theme."
...
-     `(magit-diffstat-added              ((t (:foreground ,green))))
```
Comment (wyuenho) [inline solarized.el]: `s-diffstat-added-fg`

File: solarized.el [hunk 3]
```diff
@@ -1412,25 +1370,15 @@ customize the resulting theme."
...
-     `(magit-diffstat-removed            ((t (:foreground ,red))))
```
Comment (wyuenho) [inline solarized.el]: `s-diffstat-added-bg`

File: solarized.el [hunk 4]
```diff
@@ -1412,25 +1370,15 @@ customize the resulting theme."
 ;;;;; magit
 ;;;;;; headings and diffs
      `(magit-section-highlight           ((t (:background ,base02))))
-     `(magit-section-heading             ((t (:foreground ,yellow :weight bold))))
      `(magit-section-heading-selection   ((t (:foreground ,orange :weight bold))))
-     `(magit-diff-file-heading           ((t (:weight bold))))
-     `(magit-diff-file-heading-highlight ((t (:background ,base02))))
      `(magit-diff-file-heading-selection ((t (:background ,base02
                                                           :foreground ,orange))))
-     `(magit-diff-hunk-heading
```
Comment (wyuenho) [inline solarized.el]: `s-diff-heading-[fb]g`

**Comment by thomasf** [comment]
Just started looking, just making some notes atm.

## ediff merge
### nothing seleced
#### old
image
#### new
image

### select first
#### old
image
image
image

#### new
image
image
image

### odd/even

ediff-even-diff-Ancestor / ediff-odd-diff-Ancestor are hard to see, navigation between blocks feels a bit unexpected. I wonder if they can get some tiny boost and differentiation.
#### old
image
#### new
image

**Comment by thomasf** [comment]
## smerge 

### a refined section + hl-line-mode

#### old

image

image

image

image

image

#### new

image

image

image

image

image

**Comment by thomasf** [comment]
## magit diff

### old
image

### new
image

**Comment by thomasf** [comment]
## diff-mode

The new colors are maybe a little bit to busy withot a reason. red/green foreground + hunk 
### basic diff
#### old

image

image

#### new
image

image

### marked upp a few hunks

#### old
image

image

#### new

image

image

### basic diff more color schemes

#### old

image

image


#### new

image

image

**Comment by thomasf** [comment]
it feels good after the initial look.
I will take a second look at the screenshots tomorrow or something like that.

**Comment by thomasf** [comment]
maybe something along these lines. I am removing some barley noticable new colors and making other things a little bit more coherent. (screenshots coming soon)

```diff
diff --git a/solarized.el b/solarized.el
index 65280df..40d7949 100644
--- a/solarized.el
+++ b/solarized.el
@@ -354,10 +354,6 @@ VARIANT is 'dark or 'light."
                               (solarized-color-blend "#00547f" blue 0.7)))
           (s-diff-fine-C-fg (if (eq variant 'light) "#00547f" "#d6e5e5"))
 
-          (s-diff-context-fg (if (eq variant 'light) "grey50" "grey70"))
-
-          (s-diff-heading-bg (solarized-color-blend yellow base03 0.1))
-
           (s-diffstat-added-fg green)
           (s-diffstat-changed-fg blue)
           (s-diffstat-removed-fg red)
@@ -829,16 +825,16 @@ customize the resulting theme."
        `(deadgrep-regexp-metachar-face ((,class (:inherit font-lock-constant-face))))
        `(deadgrep-search-term-face ((,class (:inherit font-lock-variable-name-face))))
 ;;;;; diff
-       `(diff-added ((,class (:background ,s-diff-B-bg :foreground ,s-diff-B-fg))))
+       `(diff-added ((,class (:background unspecified :foreground ,s-diffstat-added-fg))))
        `(diff-changed ((t nil)))
-       `(diff-removed ((,class (:background ,s-diff-A-bg :foreground ,s-diff-A-fg))))
+       `(diff-removed ((,class (:background unspecified :foreground ,s-diffstat-removed-fg))))
        `(diff-refine-added ((,class (:background ,s-diff-fine-B-bg :foreground ,s-diff-fine-B-fg))))
        `(diff-refine-changed ((,class (:background ,s-diff-fine-C-bg :foreground ,s-diff-fine-C-fg))))
        `(diff-refine-removed ((,class (:background ,s-diff-fine-A-bg :foreground ,s-diff-fine-A-fg))))
-       `(diff-header ((t (:background ,s-diff-heading-bg))))
+       `(diff-header ((t (:background nil :foreground ,base00))))
        `(diff-file-header
-         ((,class (:background ,base03 :foreground ,base0 :weight bold))))
-       `(diff-context ((,class :foreground ,s-diff-context-fg)))
+         ((,class (:background ,base03 :foreground ,base01 :weight bold))))
+       `(diff-context ((,class :foreground ,base0)))
        `(diff-indicator-added ((t (:foreground ,s-diffstat-added-fg))))
        `(diff-indicator-changed ((t (:foreground ,s-diffstat-changed-fg))))
        `(diff-indicator-removed ((t (:foreground ,s-diffstat-removed-fg))))
@@ -849,11 +845,11 @@ customize the resulting theme."
        `(diff-hl-reverted-hunk-highlight ((,class (:background ,s-diff-Ancestor-bg :foreground ,s-diff-Ancestor-fg))))
 ;;;;; ediff
        `(ediff-even-diff-A ((t (:background ,base02))))
-       `(ediff-even-diff-Ancestor ((,class (:foreground ,s-diff-context-fg))))
+       `(ediff-even-diff-Ancestor ((t (:background ,base02))))
        `(ediff-even-diff-B ((t (:background ,base02))))
        `(ediff-even-diff-C ((t (:background ,base02))))
        `(ediff-odd-diff-A ((t (:background ,base02))))
-       `(ediff-odd-diff-Ancestor ((,class (:foreground ,s-diff-context-fg))))
+       `(ediff-odd-diff-Ancestor ((t (:background ,base02))))
        `(ediff-odd-diff-B ((t (:background ,base02))))
        `(ediff-odd-diff-C ((t (:background ,base02))))
        `(ediff-current-diff-A ((,class (:background ,s-diff-A-bg :foreground ,s-diff-A-fg))))
@@ -867,7 +863,7 @@ customize the resulting theme."
 ;;;;; smerge
        `(smerge-base ((,class (:background ,s-diff-Ancestor-bg :foreground ,s-diff-Ancestor-fg))))
        `(smerge-lower ((,class (:background ,s-diff-B-bg :foreground ,s-diff-B-fg))))
-       `(smerge-markers ((t (:background ,s-diff-heading-bg))))
+       `(smerge-markers ((t (:background ,base02 :foreground ,base1))))
        `(smerge-refined-added ((,class (:background ,s-diff-fine-B-bg :foreground ,s-diff-fine-B-fg))))
        `(smerge-refined-removed ((,class (:background ,s-diff-fine-A-bg :foreground ,s-diff-fine-A-fg))))
        `(smerge-upper ((,class (:background ,s-diff-A-bg :foreground ,s-diff-A-fg))))
```

**Comment by thomasf** [comment]
##  diff-mode

diff-mode does not seem to make use of too much fancy combinations of colors so it can just be with normal background. Manually refining hunks works well.

### before
image
### after
image

**Comment by thomasf** [comment]
That was mostly it actually, the other changes are mostly about consistensy..  Please tell me if I have misunderstood something.

**Comment by wyuenho** [comment]
> The new colors are maybe a little bit to busy withot a reason. red/green foreground + a little bit more involved hunk hl should make ti clearer.

I think you probably feel this way only because you've picked some extremely busy diffs with no context to test lol.

Most people these days use the `--unified` format that's been the defacto standard for every VCS and Github and everywhere else for a few decades now, and in fact, the default in Emacs, so the scenario in your test case is very very uncommon.

Here are some screenshots with `diff -u`, `diff-font-lock-syntax` to `Don't highlight syntax` and `diff-refine`  to `Refine hunks during navigation`:

Screen Shot 2019-11-04 at 23 37 31
Screen Shot 2019-11-04 at 23 37 10

While I agree getting rid of the backgrounds in `diff-added/removed` makes a cleaner screen, I don't think this scheme works well on a light background for legibility reasons. The background and foreground colors are chosen by Magit to provide contrast, and for that purpose they work very well.

Your changes might work against a really dark background, but Solarized dark's background isn't dark enough. Besides, I'd really like to unify all the diff faces, so I'd hate to surprise someone who uses magit/diff/smerge/ediff regularly by presenting something completely different.

As to the headers and contexts, I'd like to keep them. They are subtle enough so they don't distract much, but still highlight a little bit. This is also borrowed directly from Magit.

**Comment by wyuenho** [comment]
What I do like to change tho, is the blue for diff-c in the dark themes just still seem too bright. It's a bit off, so please feel free to tune it a bit

**Comment by wyuenho** [comment]
> ediff-even-diff-Ancestor / ediff-odd-diff-Ancestor are hard to see, navigation between blocks feels a bit unexpected. I wonder if they can get some tiny boost and differentiation.

Agreed, they are too subtle to the point it's almost non-existent on the light themes. They work in dark themes tho. Let me take a look at that.

**Comment by wyuenho** [comment]
should be good now

**Comment by thomasf** [comment]
### colors
I have noticed and started working on something that *might* fix it for at least both variations of solarized and probably other color palettes as well.

I am working on completely new versions of -l and -d colors that might also work for this (because that's kind of why they were added to begin with) or I am adding another set of complementary colors. Now that we have support for one palettes per theme we can have separate color lists for the dark and bright theme so the mismatch that is currently there won't be a problem. Sadly color.el is a bit cumbersome to work with and doesn't support everything so i'm writing it using https://github.com/lucasb-eyer/go-colorful for now to not get bogged down with color space math while I experiment. I have already a good near approximation of the colors  you use that are adjusted with the same CIELAB Lightness contrast between color and background as the normal fg/bg and they look better because they are individually blended against the background for dark and light themes.

The main thing thats different from now is that the background color is blended in LAB color space which looks better and instead of blending color-fg against  normal fg the  LAB lightness is adjusted to be the same a the difference between normal fg/bg. It is promising so far, I will have to see if math can take us all the way there or not.

**Comment by thomasf** [comment]
> > ediff-even-diff-Ancestor / ediff-odd-diff-Ancestor are hard to see, navigation between blocks feels a bit unexpected. I wonder if they can get some tiny boost and differentiation.
> 
> Agreed, they are too subtle to the point it's almost non-existent on the light themes. They work in dark themes tho. Let me take a look at that.

They are also inconsistent with the other ancestor colors, like in this screenshot, if one -odd- have different colors they all should (?)

https://user-images.githubusercontent.com/330847/68092151-76951580-fe88-11e9-92db-ef2bdc016f76.png

**Comment by thomasf** [comment]
> I think you probably feel this way only because you've picked some extremely busy diffs with no context to test lol

I don't think they are particular large, not much larger than comparing feature branches to master in many projects I work on and it's a very common thing I do..  Sure, they are usually created by git and often viewed in magit so the unified format is more common. 

I don't have either `diff-font-lock-syntax` or `diff-refine`, I am not using pre release versions of emacs but it's maybe something to think about anyway. 

I will add a couple of more diff variations into https://github.com/bbatsov/solarized-emacs/tree/master/minimal-init/test-files so we have something to compare differences with. Maybe one from git diff (where -U is implied) and another with `diff -U`. I want us to work with the same examples as often as possible because it's easier to replicate for everyone. If you have suggestions for a .diff file that would be representative of some often seen scenario.

**Comment by thomasf** [comment]
While I haven't really done any real testing or tuning with the colors (now is the first time I actually even added them to the theme) here is a demo of some results from the week end.

image

image

I am just generating a bunch of colors, all of these individually for dark/bright theme because they need to blend with the background.

```
(yellow-d    . "#f2e2c1")
(yellow-hd   . "#e0c99a")
(yellow-hl   . "#977a31")
(yellow-l    . "#a08238")
(orange-d    . "#f8dac0")
(orange-hd   . "#ebb899")
(orange-hl   . "#a95431")
(orange-l    . "#b25b38")
(red-d       . "#fbd8c3")
(red-hd      . "#f1b59e")
(red-hl      . "#b4493f")
(red-l       . "#bd5146")
(magenta-d   . "#f7d9d1")
(magenta-hd  . "#eab6ba")
(magenta-hl  . "#ab4b78")
(magenta-l   . "#b45380")
(violet-d    . "#e6dedc")
(violet-hd   . "#c9c1cf")
(violet-hl   . "#616aa5")
(violet-l    . "#6972ae")
(blue-d      . "#e1e2de")
(blue-hd     . "#bfc9d4")
(blue-hl     . "#3c7caf")
(blue-l      . "#4583b7")
(cyan-d      . "#dfe6d5")
(cyan-hd     . "#bbd1c1")
(cyan-hl     . "#3a8b87")
(cyan-l      . "#42938f")
(green-d     . "#e9e5c1")
(green-hd    . "#d0ce9b")
(green-hl    . "#758532")
(green-l     . "#7d8d3a")
```

**Comment by thomasf** [comment]
the problem that causes s-diff-context-fg and gray70 to be relevant is that the highlight blocks are too bright.  I am specifically not accepting "new" base shades right now that are not generated from preexisting one.

This is however almost exactly the same so I will go with that for now: 
```
     (s-diff-context-fg (if (eq variant 'light) base0 (solarized-color-blend base1 base2 0.6)))
```

**Comment by thomasf** [comment]
the reason that s-diff-heading-bg is broken because magit-diff-hunk-heading is broken.  It doesnt work at all on dark backgrounds. I am changing it to ```(s-diff-heading-bg (if (eq variant 'light) (solarized-color-blend yellow base03 0.1) base02))``` and leaving it broken in magit since I don't have time to delve into the magit faces right now.

**Comment by thomasf** [comment]
I am pushing this to your branch and then I will merge. Regardless of any problems I still see this is much better than before.

```
diff --git a/solarized.el b/solarized.el
index 262a060..f44e3f6 100644
--- a/solarized.el
+++ b/solarized.el
@@ -353,10 +353,8 @@ VARIANT is 'dark or 'light."
                                 (solarized-color-blend "#d6e5e5" blue 0.7)
                               (solarized-color-blend "#00547f" blue 0.7)))
           (s-diff-fine-C-fg (if (eq variant 'light) "#00547f" "#d6e5e5"))
-
-          (s-diff-context-fg (if (eq variant 'light) base02 "grey70"))
-
-          (s-diff-heading-bg (solarized-color-blend yellow base03 0.1))
+          (s-diff-context-fg (if (eq variant 'light) base0 (solarized-color-blend base1 base2 0.6)))
+          (s-diff-heading-bg (if (eq variant 'light) (solarized-color-blend yellow base03 0.1) base02))
 
           (s-diffstat-added-fg green)
           (s-diffstat-changed-fg blue)
@@ -849,11 +847,11 @@ customize the resulting theme."
        `(diff-hl-reverted-hunk-highlight ((,class (:background ,s-diff-Ancestor-bg :foreground ,s-diff-Ancestor-fg))))
 ;;;;; ediff
        `(ediff-even-diff-A ((t (:background ,base02))))
-       `(ediff-even-diff-Ancestor ((,class (:foreground ,s-diff-context-fg))))
+       `(ediff-even-diff-Ancestor ((t (:background ,base02))))
        `(ediff-even-diff-B ((t (:background ,base02))))
        `(ediff-even-diff-C ((t (:background ,base02))))
        `(ediff-odd-diff-A ((t (:background ,base02))))
-       `(ediff-odd-diff-Ancestor ((,class (:foreground ,s-diff-context-fg))))
+       `(ediff-odd-diff-Ancestor ((t (:background ,base02))))
        `(ediff-odd-diff-B ((t (:background ,base02))))
        `(ediff-odd-diff-C ((t (:background ,base02))))
        `(ediff-current-diff-A ((,class (:background ,s-diff-A-bg :foreground ,s-diff-A-fg))))
```

**Pull request merged**